File: utils_android/src/main/java/org/amoustakos/utils/android/PackageManagerUtils.java
package org.amoustakos.utils.android;

import android.app.Activity;
import android.content.Context;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import timber.log.Timber;


/**
 * Utility class that provides helper methods for {@link PackageManager}.
 */
public final class PackageManagerUtils {

	/**
	 * Retrieve activities defined in the Manifest.
	 *
	 * @param context     Needed to get a {@link PackageManager} instance.
	 * @param basePackage Required to specify the manifest to look through.
	 * @param <T>         Defines the casting performed to the activities retrieved.
	 */
	public static <T extends Activity> List<Class<? extends T>> getDefinedActivities(
			@NonNull Context context,
			@NonNull String basePackage
	) {
		PackageManager pm = context.getPackageManager();
		return getDefinedActivities(pm, basePackage);
	}

	/**
	 * Retrieve activities defined in the Manifest.
	 *
	 * @param pm          The {@link PackageManager} instance to be used.
	 * @param basePackage Required to specify the manifest to look through.
	 * @param <T>         Defines the casting performed to the activities retrieved.
	 */
	@SuppressWarnings("unchecked")
	public static <T extends Activity> List<Class<? extends T>> getDefinedActivities(
			@NonNull PackageManager pm,
			@NonNull String basePackage
	) {
		List<Class<? extends T>> activities = new ArrayList<>();
		try {
			PackageInfo packageInfo = pm.getPackageInfo(basePackage, PackageManager.GET_ACTIVITIES);
			ActivityInfo[] act = packageInfo.activities;
			for (ActivityInfo ai : act) {
				try {
					if (!ai.packageName.equals(basePackage)) continue;
					Class<? extends T> cl = (Class<? extends T>) Class.forName(ai.name);
					activities.add(cl);
				} catch (ClassNotFoundException | ClassCastException cle) {
					Timber.v(cle);
				}
			}
		} catch (PackageManager.NameNotFoundException e) {
			Timber.e(e);
		}
		return activities;
	}


}

File: utils/utils_android/src/main/java/org/amoustakos/utils/android/PackageManagerUtils.kt
package org.amoustakos.utils.android

import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import timber.log.Timber
import java.util.*


/**
 * Utility class that provides helper methods for [PackageManager].
 */
object PackageManagerUtils {

    /**
     * Retrieve activities defined in the Manifest.
     *
     * @param context     Needed to get a [PackageManager] instance.
     * @param basePackage Required to specify the manifest to look through.
     * @param <T>         Defines the casting performed to the activities retrieved.
    </T> */
    fun <T : Activity> definedActivities(
            context: Context,
            basePackage: String
    ): List<Class<out T>> {
        val pm = context.packageManager
        return definedActivities(pm, basePackage)
    }

    /**
     * Retrieve activities defined in the Manifest.
     *
     * @param pm          The [PackageManager] instance to be used.
     * @param basePackage Required to specify the manifest to look through.
     * @param <T>         Defines the casting performed to the activities retrieved.
    </T> */
    fun <T : Activity> definedActivities(
            pm: PackageManager,
            basePackage: String
    ): List<Class<out T>> {
        val activities = ArrayList<Class<out T>>()
        try {
            val packageInfo = pm.getPackageInfo(basePackage, PackageManager.GET_ACTIVITIES)
            val act = packageInfo.activities
            for (ai in act) {
                try {
                    if (!ai.name.contains(basePackage)) continue
                    val cl = Class.forName(ai.name) as Class<out T>
                    activities.add(cl)
                } catch (cle: ClassNotFoundException) {
                    Timber.v(cle)
                } catch (cle: ClassCastException) {
                    Timber.v(cle)
                }

            }
        } catch (e: PackageManager.NameNotFoundException) {
            Timber.e(e)
        }

        return activities
    }


}
